hearth/lib/hearth/telemetry/otel.rb

```diff
@@ -0,0 +1,162 @@
+    # See {https://opentelemetry.io/docs/languages/ruby/exporters here} for
+    # more information.
+    #
+    # @example Exporting via console
+    #
+    #     require 'opentelemetry-sdk'
+    #
+    #     ENV['OTEL_TRACES_EXPORTER'] ||= 'console'
+    #
+    #     # configures the OpenTelemetry SDK with defaults
+    #     OpenTelemetry::SDK.configure
+    #
+    #     otel_provider = Hearth::Telemetry::OTelProvider.new
+    #     client = Service::Client.new(telemetry_provider: otel_provider)
+    class OTelProvider < TelemetryProviderBase
+      def initialize
+        unless Hearth::Telemetry.otel_loaded?
+          raise ArgumentError,
+                'Requires the `opentelemetry-sdk` gem to use OTel Provider.'
+        end
+        super(
+          # tracer_provider: OpenTelemetry.tracer_provider,
+          tracer_provider: OTelTracerProvider.new,
+          context_manager: OTelContextManager.new
+        )
+      end
+    end
+
+    # rubocop:disable Lint/MissingSuper
+
+    # OpenTelemetry-based Tracer Provider, an entry point for
+    # creating Tracer instances.
+    class OTelTracerProvider < TracerProviderBase
+      def initialize
+        @tracer_provider = OpenTelemetry.tracer_provider
+      end
+
+      def tracer(name = nil)
+        OTelTracer.new(@tracer_provider.tracer(name))
+      end
+    end
+
+    # OpenTelemetry-based Tracer, responsible for creating spans.
+    class OTelTracer < TracerBase
+      def initialize(tracer)
+        @tracer = tracer
+      end
+
+      def start_span(name, with_parent: nil, attributes: nil, kind: nil)
+        span = @tracer.start_span(
+          name,
+          with_parent: with_parent,
+          attributes: attributes,
+          kind: kind
+        )
+        OTelSpan.new(span)
+      end
+
+      def in_span(name, attributes: nil, kind: nil, &block)
+        @tracer.in_span(name, attributes: attributes, kind: kind) do |span|
+          wrapped_span = OTelSpan.new(span)
+          block.call(wrapped_span)
+        end
+      end
+    end
+
+    # OpenTelemetry-based Span, represents a single operation
+    # within a trace.
+    class OTelSpan < SpanBase
+      def initialize(span)
+        @span = span
+      end
+
+      # @api private
+      attr_reader :span
+
+      def set_attribute(key, value)
+        @span.set_attribute(key, value)
+      end
+      alias []= set_attribute
+
+      def add_attributes(attributes)
+        @span.add_attributes(attributes)
+      end
+
+      def add_event(name, attributes: nil)
+        @span.add_event(name, attributes: attributes)
+      end
+
+      def status=(status)
+        @span.status = status
+      end
+
+      def finish(end_timestamp: nil)
+        @span.finish(end_timestamp: end_timestamp)
+      end
+
+      def record_exception(exception, attributes: nil)
+        @span.record_exception(exception, attributes: attributes)
+      end
+    end
+    # rubocop:enable Lint/MissingSuper
```

Comment:
Should we just be calling super? I know it doesn't do anything on the base class, but I don't see a reason we cant call it